pydatarecognition/runcontrol.py

```diff
@@ -0,0 +1,298 @@
+            or key in self.__class__.__dict__
+        )
+
+    def __eq__(self, other):
+        if hasattr(other, "_dict"):
+            return self._dict == other._dict
+        elif isinstance(other, Mapping):
+            return self._dict == other
+        else:
+            return NotImplemented
+
+    def __ne__(self, other):
+        if hasattr(other, "_dict"):
+            return self._dict != other._dict
+        elif isinstance(other, Mapping):
+            return self._dict != other
+        else:
+            return NotImplemented
+
+    def __copy__(self):
+        return type(self)(_updaters=self._updaters, _validators=self._validators, **self._dict)
+
+    def _update(self, other):
+        """Updates the rc with values from another mapping.  If this rc has
+        if a key is in self, other, and self._updaters, then the updaters
+        value is called to perform the update.  This function should return
+        a copy to be safe and not update in-place.
+        """
+        if hasattr(other, "_dict"):
+            other = other._dict
+        elif not hasattr(other, "items"):
+            other = dict(other)
+        for k, v in other.items():
+            if v is NotSpecified:
+                pass
+            elif k in self._updaters and k in self:
+                v = self._updaters[k](getattr(self, k), v)
+            setattr(self, k, v)
+
+    def _validate(self, key, value):
+        """Validates - and possibly converts - a value based on its key and the current
+        validators.
+        """
+        validators = self._validators
+        if key in validators:
+            validator, convertor = validators[key]
+        else:
+            for vld in validators:
+                if isinstance(vld, str):
+                    continue
+                m = vld.match(key)
+                if m is not None:
+                    validator, convertor = validators[vld]
+            else:
+                validator, convertor = always_true, noop
+        return value if validator(value) else convertor(value)
+
+
+def flatten(iterable):
+    """Generator which returns flattened version of nested sequences."""
+    for el in iterable:
+        if isinstance(el, basestring):
+            yield el
+        elif isinstance(el, Iterable):
+            for subel in flatten(el):
+                yield subel
+        else:
+            yield el
+
+
+#
+# Memoization
+#
+
+
+def ishashable(x):
+    """Tests if a value is hashable."""
+    if isinstance(x, Hashable):
+        if isinstance(x, basestring):
+            return True
+        elif isinstance(x, Iterable):
+            return all(map(ishashable, x))
+        else:
+            return True
+    else:
+        return False
+
+
+DEFAULT_RC = RunControl(
+    _validators=DEFAULT_VALIDATORS,
+    builddir="_build",
+    mongodbpath=property(lambda self: os.path.join(self.builddir, "_dbpath")),
+    user_config=os.path.expanduser("~/.config/regolith/user.json"),
```

Comment:
Just a heads-up but rc gets its state from a user-specific `user.json` file here, then it looks for a `regolithrc.json` in the directory where the code is executed, then finally we can add things at runtime from the CLI argparser and so on, so it is very flexible.